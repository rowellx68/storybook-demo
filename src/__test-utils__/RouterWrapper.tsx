import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';

const RouterWrapper: React.FunctionComponent = ({ children }) => {
  return (
    <MemoryRouter keyLength={0}>
      <Route exact path='/' />
      {children}
    </MemoryRouter>
  );
};

export default RouterWrapper;
