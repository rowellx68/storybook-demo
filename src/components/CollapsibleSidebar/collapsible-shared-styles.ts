import { css } from '@emotion/core';

// import { defaultText, fontFamily, primary } from 'styles/shared-styles';
import { breakpoint } from 'styles/mixins';
import { colours, fonts } from 'styles/variables';

export const button = css({
  padding: '0.5rem',
  borderRadius: '0.25rem',
  border: 'none',
  background: '#f5f6fa',
  display: 'inline-flex',
  textDecoration: 'none',
  color: colours.gray[700],
  ':hover': {
    background: 'rgba(0, 0, 0, 0.1)',
  },
  ...breakpoint.sm({
    marginBottom: '0.5rem',
    ':last-of-type': {
      marginBottom: 0,
    },
  }),
  '&.active': {
    color: colours.primary,
    background: '#ffffff',
  },
  i: {
    fontSize: '1.5rem',
    opacity: 1,
  },
  span: {
    display: 'none',
    alignItems: 'center',
    flex: 1,
    fontSize: '1rem',
    padding: '0 0.75rem',
    fontFamily: fonts.default,
    ...breakpoint.sm({
      display: 'flex',
    }),
  },
});
