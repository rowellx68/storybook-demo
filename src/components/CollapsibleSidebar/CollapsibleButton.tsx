/** @jsx jsx */
import React from 'react';
import { css, jsx } from '@emotion/core';

import { button } from './collapsible-shared-styles';

interface ButtonProps {
  expanded: boolean;
  icon: string;
  expandIcon?: string;
  label?: string;
  onClick: () => void;
}

const buttonStyles = css(button, {
  fontSize: '1.5rem',
});

const CollapsibleButton: React.FunctionComponent<ButtonProps> = ({
  onClick,
  expanded,
  icon,
  expandIcon,
  label,
}) => {
  return (
    <button css={buttonStyles} onClick={onClick}>
      <i
        className={expanded ? expandIcon ?? icon : icon}
        aria-hidden='true'></i>
      {expanded ? <span>{label}</span> : null}
    </button>
  );
};

export default CollapsibleButton;
