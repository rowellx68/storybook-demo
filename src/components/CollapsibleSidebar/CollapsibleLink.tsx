/** @jsx jsx */
import React from 'react';
import { NavLink } from 'react-router-dom';
import { jsx } from '@emotion/core';

import { button } from './collapsible-shared-styles';

interface LinkProps {
  expanded: boolean;
  icons: string[];
  to: string;
  exact?: boolean;
  label?: string;
}

const CollapsibleLink: React.FunctionComponent<LinkProps> = ({
  expanded,
  icons,
  to,
  exact = true,
  label
}) => {
  return (
    <NavLink exact={exact} title={label} to={to} css={button}>
      {expanded ? (
        <React.Fragment>
          <i className={icons[0]} aria-hidden='true'></i>
          <span>{label}</span>
        </React.Fragment>
      ) : (
        <i className={icons[1] || icons[0]} aria-hidden='true'></i>
      )}
    </NavLink>
  );
};

export default CollapsibleLink;
