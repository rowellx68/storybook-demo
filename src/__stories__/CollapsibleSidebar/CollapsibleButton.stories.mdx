import { Meta, Story, Preview } from '@storybook/addon-docs/blocks';
import CollapsibleButton from 'components/CollapsibleSidebar/CollapsibleButton';

<Meta
  title='CollapsibleSidebar/CollapsibleButton/Docs'
  component={CollapsibleButton}
/>

# CollapsibleButton

The `CollapsibleButton` allows us to define a button with a different icons for the collapsed and expanded states. Optionally, this component also accepts a `label` to be displayed when it is expanded.

When the button is collapsed, it could look something like this:

<Story id='collapsiblesidebar-collapsiblebutton--collapsed' />

When the button is expanded, it could look something like this:

<Story id='collapsiblesidebar-collapsiblebutton--expanded' />

Both the examples above uses the ID of our existing "stories" to render them in our documentation, we could also write stories inline.

<Preview>
  <Story name='Documentation Example'>
    <div>
      <CollapsibleButton
        expanded={false}
        icon='las la-fw la-ship'
        onClick={() => null}
      />
      <br />
      <CollapsibleButton
        expanded={true}
        icon='las la-fw la-ship'
        label='Projects'
        onClick={() => null}
      />
    </div>
  </Story>
</Preview>

Using the `Preview` component for the stories gives you the ability see how the component was used! ðŸŽ‰
