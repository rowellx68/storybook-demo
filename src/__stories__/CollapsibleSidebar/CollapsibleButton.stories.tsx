import React from 'react';
import { storiesOf } from '@storybook/react';
import { action, withActions } from '@storybook/addon-actions';
import { boolean, withKnobs } from '@storybook/addon-knobs';

import { withEmotionGlobal } from '__stories__/decorators';
import 'line-awesome/dist/line-awesome/css/line-awesome.css';

import CollapsibleButton from 'components/CollapsibleSidebar/CollapsibleButton';

storiesOf('CollapsibleSidebar/CollapsibleButton', module)
  .addDecorator(withActions())
  .addDecorator(withKnobs())
  .addDecorator(withEmotionGlobal)
  .add('Collapsed', () => (
    <CollapsibleButton
      expanded={false}
      icon='las la-fw la-ship'
      onClick={action('onClick')}
    />
  ))
  .add('Expanded', () => (
    <CollapsibleButton
      expanded={true}
      icon='las la-fw la-ship'
      label='Projects'
      onClick={action('onClick')}
    />
  ))
  .add('Dynamic', () => (
    <CollapsibleButton
      expanded={boolean('Expanded', true)}
      icon='las la-fw la-ship'
      label='Projects'
      onClick={action('onClick')}
    />
  ));
