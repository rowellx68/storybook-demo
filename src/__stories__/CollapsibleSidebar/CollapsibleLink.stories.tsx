import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, withKnobs } from '@storybook/addon-knobs';

import { withEmotionGlobal, withMemoryRouter } from '__stories__/decorators';
import 'line-awesome/dist/line-awesome/css/line-awesome.css';

import CollapsibleLink from 'components/CollapsibleSidebar/CollapsibleLink';

storiesOf('CollapsibleSidebar/CollapsibleLink', module)
  .addDecorator(withKnobs())
  .addDecorator(withEmotionGlobal)
  .addDecorator(withMemoryRouter)
  .addDecorator((story) => <div>{story()}</div>);
