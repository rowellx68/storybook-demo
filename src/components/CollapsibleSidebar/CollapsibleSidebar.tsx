/** @jsx jsx */
import React, { useState } from 'react';
import { css, jsx, SerializedStyles } from '@emotion/core';

import { breakpoint } from 'styles/mixins';
import { colours } from 'styles/variables';

import CollapsibleLink from './CollapsibleLink';
import CollapsibleButton from './CollapsibleButton';

const sidebar = css({
  background: colours.light,
  padding: '0.5rem',
  display: 'flex',
  button: {
    display: 'none',
  },
  ...breakpoint.sm({
    flexDirection: 'column',
    button: {
      display: 'unset',
    },
  }),
});

const linkContainer = css({
  flex: 1,
  display: 'flex',
  justifyContent: 'space-around',
  ...breakpoint.sm({
    flexDirection: 'column',
    justifyContent: 'initial',
  }),
});

export interface CollapsibleSidebarProps {
  collapsible?: boolean;
  links: Array<{
    icons: string[];
    label?: string;
    to: string;
    exact?: boolean;
  }>;
  styles?: SerializedStyles;
}

const CollapsibleSidebar: React.FunctionComponent<CollapsibleSidebarProps> = ({
  collapsible = true,
  links = [],
  styles = {},
}) => {
  const [expanded, setExpanded] = useState(false);

  const linksStyle = css(linkContainer, {
    width: expanded ? 200 : 'auto',
  });

  const sidebarStyle = css(sidebar, styles);

  return (
    <aside css={sidebarStyle}>
      <div css={linksStyle}>
        {links.map((l, i) => (
          <CollapsibleLink
            key={i}
            expanded={expanded}
            icons={l.icons}
            label={l.label}
            to={l.to}
            exact={l.exact ?? true}
          />
        ))}
      </div>
      {collapsible ? (
        <div>
          <CollapsibleButton
            expanded={expanded}
            icon='las la-fw la-angle-double-left'
            expandIcon='las la-fw la-angle-double-right'
            onClick={() => setExpanded(!expanded)}
          />
        </div>
      ) : null}
    </aside>
  );
};

export default CollapsibleSidebar;
