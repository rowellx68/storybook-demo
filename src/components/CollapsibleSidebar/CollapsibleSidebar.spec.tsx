import React from 'react';
import { mount } from 'enzyme';

import RouterWrapper from '__test-utils__/RouterWrapper';
import CollapsibleSidebar from 'components/CollapsibleSidebar/CollapsibleSidebar';

const options = {
  wrappingComponent: RouterWrapper,
};

const links = [
  {
    icons: ['las la-fw la-home'],
    label: 'Home',
    to: '/',
  },
];

describe('CollapsibleSidebar', () => {
  it('renders aside', () => {
    const component = mount(<CollapsibleSidebar links={links} />, options);

    expect(component.exists('aside')).toEqual(true);
  });

  it('renders the correct mark-up - collapsed', () => {
    const component = mount(<CollapsibleSidebar links={links} />, options);

    expect(component).toMatchSnapshot();
  });

  it('renders link label - expanded', () => {
    const component = mount(<CollapsibleSidebar links={links} />, options);

    component.find('button').simulate('click');

    expect(component).toContainReact(<span>Home</span>);
  });

  it('renders the correct mark-up - expanded', () => {
    const component = mount(<CollapsibleSidebar links={links} />, options);

    component.find('button').simulate('click');

    expect(component).toMatchSnapshot();
  });
});
