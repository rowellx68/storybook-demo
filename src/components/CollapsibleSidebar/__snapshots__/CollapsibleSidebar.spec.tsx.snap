// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleSidebar renders the correct mark-up - collapsed 1`] = `
.emotion-6 {
  background: #f5f6fa;
  padding: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6 button {
  display: none;
}

@media screen and (min-width:640px) {
  .emotion-6 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .emotion-6 button {
    display: unset;
  }
}

.emotion-4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: auto;
}

@media screen and (min-width:640px) {
  .emotion-4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: initial;
    -webkit-justify-content: initial;
    -ms-flex-pack: initial;
    justify-content: initial;
  }
}

.emotion-0 {
  padding: 0.5rem;
  border-radius: 0.25rem;
  border: none;
  background: #f5f6fa;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #4a5568;
}

.emotion-0:hover {
  background: rgba(0,0,0,0.1);
}

@media screen and (min-width:640px) {
  .emotion-0 {
    margin-bottom: 0.5rem;
  }

  .emotion-0:last-of-type {
    margin-bottom: 0;
  }
}

.emotion-0.active {
  color: var(--primary);
  background: #ffffff;
}

.emotion-0 i {
  font-size: 1.5rem;
  opacity: 1;
}

.emotion-0 span {
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1rem;
  padding: 0 0.75rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto', 'Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans', 'Helvetica Neue',sans-serif;
}

@media screen and (min-width:640px) {
  .emotion-0 span {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

.emotion-5 {
  padding: 0.5rem;
  border-radius: 0.25rem;
  border: none;
  background: #f5f6fa;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #4a5568;
  font-size: 1.5rem;
}

.emotion-5:hover {
  background: rgba(0,0,0,0.1);
}

@media screen and (min-width:640px) {
  .emotion-5 {
    margin-bottom: 0.5rem;
  }

  .emotion-5:last-of-type {
    margin-bottom: 0;
  }
}

.emotion-5.active {
  color: var(--primary);
  background: #ffffff;
}

.emotion-5 i {
  font-size: 1.5rem;
  opacity: 1;
}

.emotion-5 span {
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1rem;
  padding: 0 0.75rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto', 'Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans', 'Helvetica Neue',sans-serif;
}

@media screen and (min-width:640px) {
  .emotion-5 span {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

<CollapsibleSidebar
  links={
    Array [
      Object {
        "icons": Array [
          "las la-fw la-home",
        ],
        "label": "Home",
        "to": "/",
      },
    ]
  }
>
  <aside
    className="emotion-6"
  >
    <div
      className="emotion-4"
    >
      <CollapsibleLink
        exact={true}
        expanded={false}
        icons={
          Array [
            "las la-fw la-home",
          ]
        }
        key="0"
        label="Home"
        to="/"
      >
        <NavLink
          className="emotion-0"
          exact={true}
          title="Home"
          to="/"
        >
          <Link
            aria-current="page"
            className="emotion-0 active"
            style={Object {}}
            title="Home"
            to={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": null,
              }
            }
          >
            <LinkAnchor
              aria-current="page"
              className="emotion-0 active"
              href="/"
              navigate={[Function]}
              style={Object {}}
              title="Home"
            >
              <a
                aria-current="page"
                className="emotion-0 active"
                href="/"
                onClick={[Function]}
                style={Object {}}
                title="Home"
              >
                <i
                  aria-hidden="true"
                  className="las la-fw la-home"
                />
              </a>
            </LinkAnchor>
          </Link>
        </NavLink>
      </CollapsibleLink>
    </div>
    <div>
      <CollapsibleButton
        expandIcon="las la-fw la-angle-double-right"
        expanded={false}
        icon="las la-fw la-angle-double-left"
        onClick={[Function]}
      >
        <button
          className="emotion-5"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="las la-fw la-angle-double-left"
          />
        </button>
      </CollapsibleButton>
    </div>
  </aside>
</CollapsibleSidebar>
`;

exports[`CollapsibleSidebar renders the correct mark-up - expanded 1`] = `
.emotion-6 {
  background: #f5f6fa;
  padding: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6 button {
  display: none;
}

@media screen and (min-width:640px) {
  .emotion-6 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .emotion-6 button {
    display: unset;
  }
}

.emotion-0 {
  padding: 0.5rem;
  border-radius: 0.25rem;
  border: none;
  background: #f5f6fa;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #4a5568;
}

.emotion-0:hover {
  background: rgba(0,0,0,0.1);
}

@media screen and (min-width:640px) {
  .emotion-0 {
    margin-bottom: 0.5rem;
  }

  .emotion-0:last-of-type {
    margin-bottom: 0;
  }
}

.emotion-0.active {
  color: var(--primary);
  background: #ffffff;
}

.emotion-0 i {
  font-size: 1.5rem;
  opacity: 1;
}

.emotion-0 span {
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1rem;
  padding: 0 0.75rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto', 'Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans', 'Helvetica Neue',sans-serif;
}

@media screen and (min-width:640px) {
  .emotion-0 span {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

.emotion-5 {
  padding: 0.5rem;
  border-radius: 0.25rem;
  border: none;
  background: #f5f6fa;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #4a5568;
  font-size: 1.5rem;
}

.emotion-5:hover {
  background: rgba(0,0,0,0.1);
}

@media screen and (min-width:640px) {
  .emotion-5 {
    margin-bottom: 0.5rem;
  }

  .emotion-5:last-of-type {
    margin-bottom: 0;
  }
}

.emotion-5.active {
  color: var(--primary);
  background: #ffffff;
}

.emotion-5 i {
  font-size: 1.5rem;
  opacity: 1;
}

.emotion-5 span {
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1rem;
  padding: 0 0.75rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','Roboto', 'Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans', 'Helvetica Neue',sans-serif;
}

@media screen and (min-width:640px) {
  .emotion-5 span {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

.emotion-4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

@media screen and (min-width:640px) {
  .emotion-4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: initial;
    -webkit-justify-content: initial;
    -ms-flex-pack: initial;
    justify-content: initial;
  }
}

<CollapsibleSidebar
  links={
    Array [
      Object {
        "icons": Array [
          "las la-fw la-home",
        ],
        "label": "Home",
        "to": "/",
      },
    ]
  }
>
  <aside
    className="emotion-6"
  >
    <div
      className="emotion-4"
    >
      <CollapsibleLink
        exact={true}
        expanded={true}
        icons={
          Array [
            "las la-fw la-home",
          ]
        }
        key="0"
        label="Home"
        to="/"
      >
        <NavLink
          className="emotion-0"
          exact={true}
          title="Home"
          to="/"
        >
          <Link
            aria-current="page"
            className="emotion-0 active"
            style={Object {}}
            title="Home"
            to={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": null,
              }
            }
          >
            <LinkAnchor
              aria-current="page"
              className="emotion-0 active"
              href="/"
              navigate={[Function]}
              style={Object {}}
              title="Home"
            >
              <a
                aria-current="page"
                className="emotion-0 active"
                href="/"
                onClick={[Function]}
                style={Object {}}
                title="Home"
              >
                <i
                  aria-hidden="true"
                  className="las la-fw la-home"
                />
                <span>
                  Home
                </span>
              </a>
            </LinkAnchor>
          </Link>
        </NavLink>
      </CollapsibleLink>
    </div>
    <div>
      <CollapsibleButton
        expandIcon="las la-fw la-angle-double-right"
        expanded={true}
        icon="las la-fw la-angle-double-left"
        onClick={[Function]}
      >
        <button
          className="emotion-5"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="las la-fw la-angle-double-right"
          />
          <span />
        </button>
      </CollapsibleButton>
    </div>
  </aside>
</CollapsibleSidebar>
`;
